### PR TITLE
Add quantization support to code generation.

### PR DESCRIPTION
**Pull request by mbartling**
At the momement, uTensor relies on quantized operators. This PR takes
that effor away from the users and puts it in the code generation.

### CODE REVIEW

File: utensor_cgen/__main__.py [hunk 1]
```diff
@@ -70,6 +70,9 @@ def _build_parser():
   parser.add_argument("--debug-comment", dest="debug_cmt",
                       action="store_true",
                       help="Add debug comments in the output source file (default: %(default)s)")
+  parser.add_argument("-Q", "--quantize", dest="quantize_flag",
+                      action="store_true", default=False
```
Comment (dboyliao) [inline utensor_cgen/__main__.py]: You missed a `,` after `default=False`.

Comment (dboyliao) [reply utensor_cgen/__main__.py]: And `default=False` is not necessary if `action="store_true"` is given.

File: utensor_cgen/pbparser.py [hunk 2]
```diff
@@ -85,6 +85,14 @@ def parse_pb(file_or_path, output_nodes=None):
   graph_def.ParseFromString(fid.read())
   fid.close()
 
+  if quantize_enabled:
+    from tensorflow.tools.graph_transforms import TransformGraph
+    if output_nodes is None:
+      target_output_nodes = []
+    # Target output nodes expects a list of targets
+    graph_def = TransformGraph(graph_def, [],
+                               target_output_nodes, ["quantize_weights", "quantize_nodes"])
```
Comment (dboyliao) [inline utensor_cgen/pbparser.py]: You need to add `"sort_by_execution_order"` to the `transforms` argument.
Or the output `graph_def` can not be imported properly with `tf.import_graph_def`

File: utensor_cgen/core.py [hunk 3]
```diff
@@ -25,6 +25,7 @@ def __init__(self, pb_file, idx_dir, embed_data_dir, method, debug_cmt=False, ou
     self.method = method
     self.debug_cmt = debug_cmt
     self.output_nodes = output_nodes
+    self.quantize_enabled = quantized_flag
```
Comment (dboyliao) [inline utensor_cgen/core.py]: In `generate`, you forget to pass `self.quantize_enabled` flag to `parse_pb` function call.

Comment (mbartling) [reply utensor_cgen/core.py]: Fixed

File: utensor_cgen/pbparser.py (hunk 2 above)
Comment (mbartling) [reply utensor_cgen/pbparser.py]: fixed

File: utensor_cgen/__main__.py (hunk 1 above)
Comment (mbartling) [reply utensor_cgen/__main__.py]: fixed

**Comment by mbartling** [comment]
@dboyliao bump

**Comment by dboyliao** [comment]
Looking good. Merging

**Comment by dboyliao** [comment]
@mbartling merge conflict in utensor_cgen/__main__.py
Please fix it first.

**Comment by mbartling** [comment]
@dboyliao fixed it. Don't know why git didnt fast forward it.

File: utensor_cgen/__main__.py [hunk 4]
```diff
@@ -71,6 +72,9 @@ def _build_parser():
   parser.add_argument("--debug-comment", dest="debug_cmt",
                       action="store_true",
                       help="Add debug comments in the output source file (default: %(default)s)")
+  parser.add_argument("-Q", "--quantize", dest="quantize_flag",
+                      action="store_true", default=False,
```
Comment (dboyliao) [inline utensor_cgen/__main__.py]: If `action` is `store_true`, the default value is automatically `False`.
The `default` keyword argument is not necessary here.
I prefer to remove it.

Comment (dboyliao) [reply utensor_cgen/__main__.py]: But if you like it this way, I'm good.

**Comment by mbartling** [comment]
@dboyliao Are you still getting that key error? I ran into the issue when not providing outputs for another PR. I think it's an issue with the parser and not the PR

**Comment by dboyliao** [comment]
But if I test it with our manually quantized pb file, it's fine.
So I'm not sure it's an issue with the parser.

**Comment by mbartling** [comment]
have you tried specifying `--outputs`? I found that fixed it for me

**Comment by dboyliao** [comment]
Do you mean `--output-nodes` ?

**Comment by dboyliao** [comment]
I think this PR is duplicate to to PR #21.
Closing it.

**Pull request closed without merging**